### PR TITLE
Feature/6139 mockserver in dev menu/implementation

### PR DESCRIPTION
**Pull request by whiskey**
## Description
Refactored environment selection and handling to support a more flexible range of environments.

The problem: defining hosts which are not 'covered' by the Telekom-issued certificate chain needed updates in code for certificate pinning and PKI.

This PR refactores the current `ServerEnvironments.json` (now `Environments.json`) to facilitate environment switches by containing additional fields for certificate hashes and public keys.

A default `Environments.json` is created on first launch. More environments can be added to that file. The identification is done via the `name` property and can be selected in the developer menu (`DEBUG` builds only) or by setting the `-env` launch argument, e.g. `-env my-env`.

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-6139

## Screenshots
Screen Shot 2021-04-20 at 18 55 01

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Client/HTTP Client/SAPDownloadedPackage.swift [hunk 1]
```diff
@@ -34,48 +35,6 @@ struct SAPDownloadedPackage: Fingerprinting {
...
-				}
-			}
-			
-			return false
-		}
-
-		func callAsFunction(_ package: SAPDownloadedPackage) -> Bool {
-			verify(package)
-		}
-	}
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Client/HTTP Client/SAPDownloadedPackage.swift:78]: Totally redundant and never used…

File: src/xcode/ENA/ENA/Source/Client/HTTP Client/CachingHTTPClient.swift [hunk 2]
```diff
@@ -33,12 +33,12 @@ class CachingHTTPClient: AppConfigurationFetching, StatisticsFetching, QRCodePos
 	///   - session: An optional session to use for network requests. Default is based on a predefined configuration.
 	///   - signatureVerifier: The signatureVerifier to use for package validation.
 	init(
-		serverEnvironmentProvider: ServerEnvironmentProviding,
+		environmentProvider: EnvironmentProviding = Environments(),
```
Comment (SwissMountainFrog) [inline src/xcode/ENA/ENA/Source/Client/HTTP Client/CachingHTTPClient.swift:36]: Just thinking aloud. You have refactored the `serverEnvironmentProvider` to `environmentProvider` but the implementation, later on, is `Environments()`. Wouldn't it be semantically better to refactor it to `environmentsProvider` (plural)?

The same goes with the protocol `EnvironmentProviding` could be semantically better as `EnvironmentsProviding`.

File: src/xcode/ENA/ENA/Source/Client/HTTP Client/SAPDownloadedPackage.swift [hunk 3]
```diff
@@ -22,6 +22,7 @@ struct SAPDownloadedPackage: Fingerprinting {
 		do {
 			self = try archive.extractKeyPackage()
 		} catch {
+			Log.error("Could not extract key package: \(error.localizedDescription)", log: .localData, error: error)
```
Comment (SwissMountainFrog) [inline src/xcode/ENA/ENA/Source/Client/HTTP Client/SAPDownloadedPackage.swift:25]: 👍 👍 👍

File: src/xcode/ENA/ENA/Source/Environment/__tests__/EnvironmentTests.swift [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+	func testSelectedEnvironment() {
+		let testEnvironments = loadTestEnvironment()
+		// valid selection
+		["prod", "TestEnvironment0", "TestEnvironment1", "TestEnvironment2"].shuffled().forEach { name in
+			UserDefaults.standard.setValue(name, forKey: Environments.selectedEnvironmentKey)
+			XCTAssertEqual(testEnvironments.currentEnvironment().name, name)
+		}
+
+		// invalid selection will result in a `fatalError` so no need (and chance) to test it
```
Comment (SwissMountainFrog) [inline src/xcode/ENA/ENA/Source/Environment/__tests__/EnvironmentTests.swift:42]: 😄  👍

**Review by SwissMountainFrog — CHANGES_REQUESTED** [review]

**Review by SwissMountainFrog — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Client/HTTP Client/SAPDownloadedPackage.swift (hunk 1 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Client/HTTP Client/SAPDownloadedPackage.swift:78]: Seems you found the code I missed to remove 😅

File: src/xcode/ENA/ENA/Source/Client/Security/iOS 12 Fallback/CryptoKitFallbacks.swift [hunk 5]
```diff
@@ -170,14 +170,21 @@ protocol PublicKeyProtocol {
...
 	let x963Representation: Data
 
 	/// Initializes a PublicKey from a given key string.
 	/// - Parameters:
 	///   - pkString: A string representation of the public key to store
-	init(with pkString: StaticString) {
-		let rawData = Data(staticBase64Encoded: pkString)
+	init(with pkString: String) throws {
+		guard let rawData = Data(base64Encoded: pkString) else {
+			throw Error.publicKeyInitFailed
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Client/Security/iOS 12 Fallback/CryptoKitFallbacks.swift:186]: should we add here a log?

File: src/xcode/ENA/ENA/Source/Extensions/URLSession+Default.swift [hunk 6]
```diff
@@ -7,13 +7,12 @@ import Foundation
...
 	class func coronaWarnSession() -> URLSession {
 		#if DISABLE_CERTIFICATE_PINNING
-
 		/// Disable certificate pinning while app is running on:
 		/// Community, Debug, TestFlight, UITesting modes
 		let coronaWarnURLSessionDelegate: CoronaWarnURLSessionDelegate? = nil
 		#else
 		let coronaWarnURLSessionDelegate = CoronaWarnURLSessionDelegate(
-			publicKeyHash: "f30c3959de6b062374f037c505fb3864e1b0678086252ab457ddd97c729d06ab"
+			publicKeyHash: Environments().currentEnvironment().pinningKeyHash
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Extensions/URLSession+Default.swift:15]: 💎

**Review by pascal-brause — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Client/Security/iOS 12 Fallback/CryptoKitFallbacks.swift (hunk 5 above)
Comment (whiskey) [reply src/xcode/ENA/ENA/Source/Client/Security/iOS 12 Fallback/CryptoKitFallbacks.swift:186]: I normally don't log errors on throw but rather on catch. In previous projects this ended up in double logging issues.

**Pull request merged**